Changes to Actions::UpdateU

nilsdeppe opened the pull request:
## Proposed changes

- UpdateU no longer removes dt_vars from the DataBox. Basically we don't want to have to re-allocate every time step.
- UpdateU uses `Tags::dt<system::variables_tag>` instead of `system::dt_variables_tag`.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`


### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

nilsdeppe committed:
UpdateU no longer removes dt_vars from box

The temporal derivative is no longer removed from the box so that it
can be mutated later. The reason for this change is that later we will
re-use the oldest history data in the time stepper history to avoid
having to re-allocate data the temporal derivatives. Currently the
time steppers do not support plundering the history in such a fashion
so the allocation is currently still necessary.

nilsdeppe committed:
UpdateU wraps system::variables_tag in Tags::dt

This means systems no longer have to have a dt_variables_tag type alias.

wthrowe commented on src/Time/Actions/UpdateU.hpp:
```diff
@@ -55,16 +56,14 @@ struct UpdateU {
               Parallel::get<CacheTags::TimeStepper>(cache);
 
-          history.emplace_back(time, vars, std::move(dt_vars));
+          history.emplace_back(time, vars, dt_vars);
```
Doesn't this make things worse?  You've replaced an allocation for the DataBox with an allocation for the history and a copy into said allocation, so we seem to have no net change in allocations but an extra copy.

nilsdeppe replied:
Yes, if I remember correctly you and I had talked about re-using the history data at some point. I didn't want to make changes to that because I know you're still working on how the time stepper interactions with the DataBox a bit. My plan is to first not remove the dt_variables from the DataBox, then later be able to re-use the oldest entry in the history data. Is there currently a way to steal the oldest entry?

wthrowe replied:
Yes, reusing history allocations has been discussed, and still sounds like a good idea.  It shouldn't be done right now, because pull request #469 completely redoes the history structures.  (The new versions do not yet support this optimization, but I intend to add it after we get the initial version agreed upon.)  This is probably the direction UpdateU will have to go in for that, so I'm not strongly opposed to this.

I do think that the commit message needs to not claim this is saving on allocations.

nilsdeppe replied:
Sounds good to me. I think it would be useful for the commit message to actually explain the direction this should head. That is, make it clear there currently isn't a reduction in allocations but that this is a step in that direction and that future changes will avoid the allocation. Does that sound good to you?

wthrowe replied:
Yeah, explaining the motivation is good.

nilsdeppe replied:
Okay, I'll push an update. Thanks!

nilsdeppe commented:
Okay, I rebased and updated the commit message. @wthrowe let me know if you're now okay with these changes.

@kidder can you look over this?

Since there are no doc changes I haven't pushed documentation.

nilsdeppe commented:
Rebased this on top of newest develop to resolve merge conflicts. @wthrowe are you happy with the updated commit message?

wthrowe reviewed (dismissed):
I'll approve it, but Travis is failing so you're going to have to rebase anyway.

nilsdeppe commented:
Fixed warning with GCC and rebased on develop
